kyuubi-server/src/main/scala/org/apache/kyuubi/engine/YarnApplicationOperation.scala

```diff
@@ -95,74 +95,139 @@ class YarnApplicationOperation extends ApplicationOperation with Logging {
+      case Some(sessionConf) =>
+        (proxyUser) match {
+          case Some(user) =>
+            Utils.doAs(user) { () =>
+              var yarnClient: YarnClient = null
+              try {
+                val yarnConf = KyuubiHadoopUtils.newYarnConfiguration(sessionConf)
+                yarnClient = createYarnClient(yarnConf)
+                action(yarnClient)
+              } finally {
+                Utils.tryLogNonFatalError(yarnClient.close())
+              }
+            }
+          case None =>
+            var yarnClient: YarnClient = null
+            try {
+              yarnClient = createAdminYarnClient(sessionConf).get
+              action(yarnClient)
+            } finally {
+              Utils.tryLogNonFatalError(yarnClient.close())
+            }
+        }
+      case None =>
+        withYarnClient(proxyUser)(action)
+    }
+  }
+
+  def createAdminYarnClient(conf: KyuubiConf): Option[YarnClient] = {
+
+    val yarnConf = KyuubiHadoopUtils.newYarnConfiguration(conf)
+
+    def createYarnClientWithCurrentUser(): Option[YarnClient] = {
+      val c = createYarnClient(yarnConf)
+      info(s"Creating admin YARN client with current user: ${Utils.currentUser}.")
+      Some(c);
+    }
+
+    def createYarnClientWithProxyUser(proxyUser: String): Option[YarnClient] =
+      Utils.doAs(proxyUser) { () =>
+        val c = createYarnClient(yarnConf)
+        info(s"Creating admin YARN client with proxy user: $proxyUser.")
+        Some(c);
+      }
+
+    YarnUserStrategy.withName(conf.get(KyuubiConf.YARN_USER_STRATEGY)) match {
+      case NONE =>
+        createYarnClientWithCurrentUser()
+      case ADMIN if conf.get(KyuubiConf.YARN_USER_ADMIN) == Utils.currentUser =>
+        createYarnClientWithCurrentUser()
+      case ADMIN =>
+        createYarnClientWithProxyUser(conf.get(KyuubiConf.YARN_USER_ADMIN))
+      case OWNER =>
+        info("Skip initializing admin YARN client")
+        None
+    }
+  }
+
   override def isSupported(appMgrInfo: ApplicationManagerInfo): Boolean =
     appMgrInfo.resourceManager.exists(_.toLowerCase(Locale.ROOT).startsWith("yarn"))
 
   override def killApplicationByTag(
+      sessionConf: Option[KyuubiConf],
       appMgrInfo: ApplicationManagerInfo,
       tag: String,
-      proxyUser: Option[String] = None): KillResponse = withYarnClient(proxyUser) { yarnClient =>
-    try {
-      val reports = yarnClient.getApplications(null, null, Set(tag).asJava)
-      if (reports.isEmpty) {
-        (false, NOT_FOUND)
-      } else {
-        try {
-          val applicationId = reports.get(0).getApplicationId
-          yarnClient.killApplication(applicationId)
-          (true, s"Succeeded to terminate: $applicationId with $tag")
-        } catch {
-          case e: Exception =>
-            (false, s"Failed to terminate application with $tag, due to ${e.getMessage}")
+      proxyUser: Option[String] = None): KillResponse =
+    withNewYarnClient(proxyUser, sessionConf) { yarnClient =>
```

Comment:
It seems that `newYarnClient` is not always needed. Can we use it only when needed? Like, the `hadoop_conf_dir/yarn_conf_dir` engine evn exists in `sessionConf`.

Reply:
Thanks for your help, you're right, I'll make the changes.